lib/block_view_atomic_txns.go

```diff
@@ -0,0 +1,472 @@
+	if len(txn.TxOutputs) != 0 {
+		return RuleErrorAtomicTxnsWrapperMustHaveZeroOutputs
+	}
+
+	// There exists three design options for txn.TxnFeeNanos rules in atomic transaction wrappers:
+	//	(1) Force txn.TxnFeeNanos to equal zero.
+	//	(2) Force txn.TxnFeeNanos to equal the sum of the internal transaction's txn.TxnFeeNanos fields.
+	//	(3) Ignore txn.TxnFeeNanos entirely.
+	//
+	// Because txn.TxnFeeNanos gets used in several places for non-connection logic (e.g. BMF),
+	// it's important to use design option (2) to be consistent across core. This check as a result
+	// becomes extremely important in _connectAtomicTxnsWrapper().
+	var totalInnerTxnFees uint64
+	var err error
+	for _, innerTxn := range txn.TxnMeta.(*AtomicTxnsWrapperMetadata).Txns {
+		totalInnerTxnFees, err = SafeUint64().Add(totalInnerTxnFees, innerTxn.TxnFeeNanos)
+		if err != nil {
+			return RuleErrorAtomicTxnsWrapperHasInternalFeeOverflow
+		}
+	}
+	if txn.TxnFeeNanos != totalInnerTxnFees {
+		return RuleErrorAtomicTxnsWrapperMustHaveEqualFeeToInternalTxns
+	}
+
+	// Technically the txn.TxnNonce field could be anything but for consistent
+	// hashing we force all fields of the nonce to be zero. This also makes
+	// it more consistent with the rest of the rules regarding atomic transactions
+	// wrappers.
+	if txn.TxnNonce.ExpirationBlockHeight != 0 || txn.TxnNonce.PartialID != 0 {
+		return RuleErrorAtomicTxnsWrapperMustHaveZeroedNonce
+	}
+
+	// NOTE: We do not enforce rules on txn.ExtraData as it's both useful
+	// 		 for app developers and is being paid for via txn.TxnFeeNanos.
+
+	return nil
+}
+
+func _verifyAtomicTxnsChain(txnMeta *AtomicTxnsWrapperMetadata) error {
+	// Validate:
+	//	(1) The inner transactions are not additional redundant atomic transactions wrappers.
+	//  (2) The inner transactions are meant to be included in an atomic transaction.
+	//	(3) The start point is the first inner transaction and there's only one start point.
+	// We also collect the atomic hash of each inner transaction here for convenience.
+	var atomicHashes []*BlockHash
+	for ii, innerTxn := range txnMeta.Txns {
+		// Validate this transaction is not another redundant atomic transaction.
+		if innerTxn.TxnMeta.GetTxnType() == TxnTypeAtomicTxnsWrapper {
+			return RuleErrorAtomicTxnsHasAtomicTxnsInnerTxn
+		}
+
+		// Validate the inner transaction as meant to be included in an atomic transaction.
+		if !innerTxn.IsAtomicTxnsInnerTxn() {
+			return RuleErrorAtomicTxnsHasNonAtomicInnerTxn
+		}
+
+		// Validate the starting point of the atomic transactions chain.
+		_, keyExists := innerTxn.ExtraData[AtomicTxnsChainLength]
+		if !keyExists && ii == 0 {
+			return RuleErrorAtomicTxnsMustStartWithChainLength
+		}
+		if keyExists && ii > 0 {
+			return RuleErrorAtomicTxnsHasMoreThanOneStartPoint
+		}
+
+		// The error check in AtomicHash() is almost redundant, but we must keep it in the event
```

Comment:
well we actually do this to be more efficient when checking the atomic hashes below as well! This allows us to avoid calling `AtomicHash` an extra time below.